std/protocols/permutation.asm

```diff
@@ -1,44 +1,152 @@
+};
+
+/// Maps [x_1, x_2, ..., x_n] to alpha**(n - 1) * x_1 + alpha ** (n - 2) * x_2 + ... + x_n
+let<T: Add + Mul + FromLiteral> compress_expression_array: T[], Fp2<T> -> Fp2<T> = |expr_array, alpha| fold(
+    expr_array,
+    from_base(0),
+    |sum_acc, el| add_ext(mul_ext(alpha, sum_acc), from_base(el))
+);
+
+/// Takes a selector (0 or 1) and a value, returns value (if selector == 1) or 1 (if selector == 0)
+/// Implemented as: selector * (value - 1) + 1
+let<T: Add + Mul + Sub + FromLiteral> selected_or_one: T, Fp2<T> -> Fp2<T> = |selector, value| add_ext(mul_ext(from_base(selector), sub_ext(value, from_base(1))), from_base(1));
+
+// Compute z' = z * selected_or_one(sel_a, beta - a) / selected_or_one(sel_b, beta - b), using extension field arithmetic
+// This is intended to be used as a hint in the extension field case; for the base case
+// automatic witgen is smart enough to figure out the value if the accumulator.
+let compute_next_z: Fp2<expr>, Constr -> fe[] = query |acc, permutation_constraint| {
+
+    let (lhs_selector, lhs, rhs_selector, rhs) = unpack_permutation_constraint(permutation_constraint);
+
+    let alpha = if len(lhs) > 1 {
+        Fp2::Fp2(alpha1, alpha2)
+    } else {
+        // The optimizer will have removed alpha, but the compression function
+        // still accesses it (to multiply by 0 in this case)
+        from_base(0)
+    };
+    let beta = Fp2::Fp2(beta1, beta2);
+    
+    let lhs_folded = selected_or_one(lhs_selector, sub_ext(beta, compress_expression_array(lhs, alpha)));
+    let rhs_folded = selected_or_one(rhs_selector, sub_ext(beta, compress_expression_array(rhs, alpha)));
+    
+    // acc' = acc * lhs_folded / rhs_folded
+    let res = mul_ext(
+        eval_ext(mul_ext(acc, lhs_folded)),
+        inv_ext(eval_ext(rhs_folded))
+    );
+
+    match res {
+        Fp2::Fp2(a0_fe, a1_fe) => [a0_fe, a1_fe]
+    }
+};
 
 // Adds constraints that enforce that lhs is a permutation of rhs
 // Arguments:
-// - acc: A phase-2 witness column to be used as the accumulator
-// - lhs_selector: (assumed to be) binary selector to check which elements from the LHS to include
-// - lhs: An array of expressions
-// - rhs_selector: (assumed to be) binary selector to check which elements from the RHS to include
-// - rhs: An array of expressions
-let permutation = |acc, lhs_selector, lhs, rhs_selector, rhs| {
+// - acc: A phase-2 witness column to be used as the accumulator. If 2 are provided, computations
+//        are done on the F_{p^2} extension field.
+// - permutation_constraint: The permutation constraint
+let permutation: expr[], Constr -> Constr[] = |acc, permutation_constraint| {
+
+    let (lhs_selector, lhs, rhs_selector, rhs) = unpack_permutation_constraint(permutation_constraint);
 
     let _ = assert(len(lhs) == len(rhs), || "LHS and RHS should have equal length");
-    let _ = assert(modulus() > 2**100, || "This implementation assumes a large field");
+    let with_extension = match len(acc) {
```

Comment:
`let permutation_with_extension: expr[], Constr -> Constr[] = |acc, permutation_constraint| {
    let _ = assert(len(acc) == 2, || "Accumulators should have a lenght of 2 to use the extension field");

    let (lhs_selector, lhs, rhs_selector, rhs) = unpack_permutation_constraint(permutation_constraint);

    let _ = assert(len(lhs) == len(rhs), || "LHS and RHS should have equal length");

    let fp2_from_array = |arr| Fp2::Fp2(arr[0], arr[1]);
    let acc_ext = fp2_from_array(acc);
    let alpha = fp2_from_array([alpha1, alpha2]);
    let beta = fp2_from_array([beta1, beta2]);

    let lhs_folded = selected_or_one(lhs_selector, sub_ext(beta, compress_expression_array(lhs, alpha)));
    let rhs_folded = selected_or_one(rhs_selector, sub_ext(beta, compress_expression_array(rhs, alpha)));

    let next_acc = next_ext(acc_ext);

    let (update_expr_1, update_expr_2) = unpack_ext(sub_ext(
        mul_ext(rhs_folded, next_acc),
        mul_ext(lhs_folded, acc_ext)
    ));

    let (acc_1, acc_2) = unpack_ext(acc_ext);

    [
        // First and last z needs to be 1
        // (because of wrapping, the z[0] and z[N] are the same)
        is_first * (acc_1 - 1) = 0,

        // Note that if with_extension is false, this generates 0 = 0 and is removed
        // by the optimizer.
        is_first * acc_2 = 0,

        // Assert that the update rule has been obeyed
        update_expr_1 = 0,

        // Again, update_expr_2 will be equal to 0 in the non-extension case.
        update_expr_2 = 0
    ]
};`

Reply:
`let permutation_with_extension: expr[], Constr -> Constr[] = |acc, permutation_constraint| {
    let _ = assert(len(acc) == 2, || "Accumulators should have a lenght of 2 to use the extension field");

    let (lhs_selector, lhs, rhs_selector, rhs) = unpack_permutation_constraint(permutation_constraint);

    let _ = assert(len(lhs) == len(rhs), || "LHS and RHS should have equal length");

    let fp2_from_array = |arr| Fp2::Fp2(arr[0], arr[1]);
    let acc_ext = fp2_from_array(acc);
    let alpha = fp2_from_array([alpha1, alpha2]);
    let beta = fp2_from_array([beta1, beta2]);

    let lhs_folded = selected_or_one(lhs_selector, sub_ext(beta, compress_expression_array(lhs, alpha)));
    let rhs_folded = selected_or_one(rhs_selector, sub_ext(beta, compress_expression_array(rhs, alpha)));

    let next_acc = next_ext(acc_ext);

    let (update_expr_1, update_expr_2) = unpack_ext(sub_ext(
        mul_ext(rhs_folded, next_acc),
        mul_ext(lhs_folded, acc_ext)
    ));

    let (acc_1, acc_2) = unpack_ext(acc_ext);

    [
        // First and last z needs to be 1
        // (because of wrapping, the z[0] and z[N] are the same)
        is_first * (acc_1 - 1) = 0,

        // Note that if with_extension is false, this generates 0 = 0 and is removed
        // by the optimizer.
        is_first * acc_2 = 0,

        // Assert that the update rule has been obeyed
        update_expr_1 = 0,

        // Again, update_expr_2 will be equal to 0 in the non-extension case.
        update_expr_2 = 0
    ]
};`

Reply:
Roughly this:
```
let permutation = |acc, lhs_selector, lhs, rhs_selector, rhs| {

    let _ = assert(len(lhs) == len(rhs), || "LHS and RHS should have equal length");
    let _ = assert(modulus() > 2**100, || "This implementation assumes a large field");

    // If the selector is 1, contribute a factor of `beta - compress_expression_array(lhs)` to accumulator.
    // If the selector is 0, contribute a factor of 1 to the accumulator. 
    let lhs_folded = lhs_selector * (beta - compress_expression_array(lhs) - 1) + 1;
    let rhs_folded = rhs_selector * (beta - compress_expression_array(rhs) - 1) + 1;

    [
        // First and last z needs to be 1
        // (because of wrapping, the z[0] and z[N] are the same)
        is_first * (acc - 1) = 0,

        // Update rule:
        // acc' = acc * lhs_folded / rhs_folded
        rhs_folded * acc' = acc * lhs_folded
    ]
};
```
```
let permutation_with_extension: expr[], Constr -> Constr[] = |acc, permutation_constraint| {
    let _ = assert(len(acc) == 2, || "Accumulators should have a lenght of 2 to use the extension field");

    let (lhs_selector, lhs, rhs_selector, rhs) = unpack_permutation_constraint(permutation_constraint);

    let _ = assert(len(lhs) == len(rhs), || "LHS and RHS should have equal length");

    let fp2_from_array = |arr| Fp2::Fp2(arr[0], arr[1]);
    let acc_ext = fp2_from_array(acc);
    let alpha = fp2_from_array([alpha1, alpha2]);
    let beta = fp2_from_array([beta1, beta2]);

    let lhs_folded = selected_or_one(lhs_selector, sub_ext(beta, compress_expression_array(lhs, alpha)));
    let rhs_folded = selected_or_one(rhs_selector, sub_ext(beta, compress_expression_array(rhs, alpha)));

    let next_acc = next_ext(acc_ext);

    let (update_expr_1, update_expr_2) = unpack_ext(sub_ext(
        mul_ext(rhs_folded, next_acc),
        mul_ext(lhs_folded, acc_ext)
    ));

    let (acc_1, acc_2) = unpack_ext(acc_ext);

    [
        // First and last z needs to be 1
        // (because of wrapping, the z[0] and z[N] are the same)
        is_first * (acc_1 - 1) = 0,

        // Note that if with_extension is false, this generates 0 = 0 and is removed
        // by the optimizer.
        is_first * acc_2 = 0,

        // Assert that the update rule has been obeyed
        update_expr_1 = 0,

        // Again, update_expr_2 will be equal to 0 in the non-extension case.
        update_expr_2 = 0
    ]
}; 
```

Reply:
Does an "if statement" have a computational overhead for the Executor or Witgen?

Reply:
No, all this code is "compile-time" code.